deepmd/pt/infer/deep_eval.py

```diff
@@ -195,54 +303,33 @@ def _eval_model(
         )
         if isinstance(batch_output, tuple):
             batch_output = batch_output[0]
-        energy_out = batch_output["energy"].reshape(nframes, 1).detach().cpu().numpy()
-        if "atom_energy" in batch_output:
-            atomic_energy_out = (
-                batch_output["atom_energy"]
-                .reshape(nframes, natoms, 1)
-                .detach()
-                .cpu()
-                .numpy()
-            )
-        force_out = (
-            batch_output["force"].reshape(nframes, natoms, 3).detach().cpu().numpy()
-        )
-        virial_out = batch_output["virial"].reshape(nframes, 9).detach().cpu().numpy()
-        if "atomic_virial" in batch_output:
-            atomic_virial_out = (
-                batch_output["atomic_virial"]
-                .reshape(nframes, natoms, 9)
-                .detach()
-                .cpu()
-                .numpy()
-            )
 
-        if not atomic:
-            return energy_out, force_out, virial_out
-        else:
-            return (
-                energy_out,
-                force_out,
-                virial_out,
-                atomic_energy_out,
-                atomic_virial_out,
-            )
-
-    def get_ntypes(self) -> int:
-        """Get the number of atom types of this model."""
-        return len(self.type_map)
-
-    def get_type_map(self) -> List[str]:
-        """Get the type map (element name of the atom types) of this model."""
-        return self.type_map
+        results = []
+        for ii, odef in enumerate(self.output_def.var_defs.values()):
+            if not atomic and "_redu" not in odef.name and "_derv_r" not in odef.name:
+                continue
+            pt_name = self._OUTDEF_DP2PT[odef.name]
+            if pt_name in batch_output:
+                shape = self._get_output_shape(odef.name, nframes, natoms, odef.shape)
+                out = batch_output[pt_name].reshape(shape).detach().cpu().numpy()
+                results.append(out)
+        return tuple(results)
 
-    def get_dim_fparam(self) -> int:
-        """Get the number (dimension) of frame parameters of this DP."""
-        return 0
-
-    def get_dim_aparam(self) -> int:
-        """Get the number (dimension) of atomic parameters of this DP."""
-        return 0
+    def _get_output_shape(self, name, nframes, natoms, shape):
+        if "_redu" in name:
+            if "_derv_c" in name:
+                return [nframes, *shape[:-2], 9]
+            else:
+                return [nframes, *shape, 1]
+        else:
+            if "_derv_c" in name:
+                return [nframes, *shape[:-2], natoms, 9]
+            elif "_derv_r" in name:
+                return [nframes, *shape[:-1], natoms, 3]
+            else:
+                # Something wrong here?
+                # return [nframes, *shape, natoms, 1]
+                return [nframes, natoms, *shape, 1]
```

Comment:
could you fix the wrong shape in a new PR?

Reply:
The old TF UT failed if the shape is `[nframes, *shape, natoms, 1]`. Currently, I am not sure what the correct result is, i.e. whether the result matches its shape.